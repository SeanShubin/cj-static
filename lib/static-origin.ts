import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as cloudfront from '@aws-cdk/aws-cloudfront';
import { ManagedPolicy, PolicyStatement, Effect } from '@aws-cdk/aws-iam';
export interface IStaticOrigin {

    readonly store: s3.Bucket,
    readonly accessId: cloudfront.OriginAccessIdentity,
    readonly writeAccessPolicy: ManagedPolicy,

}


export class StaticOrigin extends cdk.Construct implements IStaticOrigin {
    writeAccessPolicy: ManagedPolicy;
    store: s3.Bucket;
    accessId: cloudfront.OriginAccessIdentity;

    constructor(scope: cdk.Construct) {
        super(scope, "Origin");

        this.store = new s3.Bucket(this, 'Content');

        this.accessId = new cloudfront.OriginAccessIdentity(this, 'access-id');

        this.writeAccessPolicy = new WriteAccessPolicy(this, this.store);
        new AccessIdReadPolicy(this, this.store, this.accessId);

    }
}


class AccessIdBucketReadAnyStatement extends PolicyStatement {
    constructor(bucket: s3.Bucket, accessId: cloudfront.OriginAccessIdentity) {
        super();
        this.effect = Effect.ALLOW;
        this.addActions('s3:GetObject');
        this.addResources(bucket.arnForObjects('*'));
        this.addCanonicalUserPrincipal(accessId.cloudFrontOriginAccessIdentityS3CanonicalUserId);
    }
}

class BucketWriteAnyStatement extends PolicyStatement {
    constructor(bucket: s3.Bucket) {
        super();
        this.effect = Effect.ALLOW;
        this.addActions(
            's3:PutObject',
            's3:DeleteObject',
            's3:GetObject',
            's3:ListBucket',
            's3:AbortMultipartUpload',
            's3:ListBucketMultipartUploads',
            's3:ListMultipartUploadParts');
        this.addResources(bucket.arnForObjects('*'), bucket.bucketArn);
    }
}


export class AccessIdReadPolicy extends s3.BucketPolicy {
    bucket: s3.Bucket;
    accessId: cloudfront.OriginAccessIdentity;

    constructor(scope: cdk.Construct, bucket: s3.Bucket, accessId: cloudfront.OriginAccessIdentity) {
        super(scope, bucket.node.id + "AccessIdReadAccess", {
            bucket: bucket
        });
        this.document.addStatements(new AccessIdBucketReadAnyStatement(bucket, accessId));
        this.bucket = bucket;
        this.accessId = accessId;
    }
}

export class WriteAccessPolicy extends ManagedPolicy {

    bucket: s3.Bucket;

    constructor(scope: cdk.Construct, bucket: s3.Bucket) {
        super(scope, bucket.node.id + "WriteAccess", {
            statements: [new BucketWriteAnyStatement(bucket)]
        });
        this.bucket = bucket;
    }
}


