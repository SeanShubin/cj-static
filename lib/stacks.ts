import cdk = require('@aws-cdk/core');

import {StaticOrigin} from './static-origin';


function asName(s: string): string {
    return s.replace(/[^A-Za-z0-9-]/g, '-');
}

export class DataStack extends cdk.Stack {
    constructor(scope: cdk.Construct, props?: cdk.StackProps) {
        super(scope, "StaticData", props);

        new StaticOrigin(this, "Origin");

      }
}

export class CertificateStack extends cdk.Stack {
    constructor(scope: cdk.Construct, domain: string, props?: cdk.StackProps) {
        super(scope, "Certificate-" + asName(domain), props);

      }
}

export class DistributionStack extends cdk.Stack {
    constructor(scope: cdk.Construct, domain: string, props?: cdk.StackProps) {
        super(scope, "Distribution-" + asName(domain), props);

      }
}
