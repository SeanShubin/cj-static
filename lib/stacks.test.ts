import { StaticOrigin } from "./static-origin";
import { CfnOutput, App, Stack } from "@aws-cdk/core";
import { DnsValidatedCertificate, ValidationMethod } from '@aws-cdk/aws-certificatemanager';
import { DataStack, CertificateStack, DistributionStack } from "./stacks";
import { expect as expectCDK, haveResource, haveResourceLike, SynthUtils } from '@aws-cdk/assert';
import { Bucket } from "@aws-cdk/aws-s3";

test('Data Stack Has Content Write Access Policy Arn Output', () => {

    // when
    const stack = new DataStack(new App());

    // then
    const output = stack.node.children.filter(c => c instanceof CfnOutput && c.node.id === 'WriteContentPolicyArn');
    expect(output.length).toBe(1);

    // snapshot on the output configuration
    expect(SynthUtils.toCloudFormation(stack).Outputs.WriteContentPolicyArn).toMatchSnapshot();

});

test('Certificate Stack Has Certificate', () => {


    const domain = "foo.bar.com";
    const stack = new CertificateStack(new App, domain);

    // then
    expectCDK(stack).to(haveResource('AWS::CertificateManager::Certificate', {
            DomainName: domain,
            ValidationMethod:  ValidationMethod.DNS,
        }));


});

test('Certificate Stack Deploys to US East 1', () => {

    const stack = new CertificateStack(new App, "foo.bar.com");

    expect(stack.region).toBe("us-east-1");
});

function squish(s: string): string {
    return s.replace(/[^a-zA-Z0-9]/g, '');
}

test('Certificate Stack Has Certificate Arn Output', () => {

    // when
    const stack = new CertificateStack(new App, "foo.bar.com");

    // then
    const nodeId = "Certificate-foo-bar-com-Arn";
    const output = stack.node.children.filter(c => c instanceof CfnOutput && c.node.id === nodeId);
    expect(output.length).toBe(1);

    // snapshot the output configuration
    const tpl = SynthUtils.toCloudFormation(stack);
    expect(tpl.Outputs[squish(nodeId)]).toMatchSnapshot();

});

test('Distribution stack has distribution', () => {
    // given
    const app = new App;
    const bstack = new Stack(app, "data", {});
    const origin = new Bucket(bstack, "o");

    // when
    const stack = new DistributionStack(app,  "bar.foo.com", origin);

    // then
    expectCDK(stack).to(haveResourceLike('AWS::CloudFront::Distribution', {
        DistributionConfig: {
            Origins: [{
                OriginPath: "/bar.foo.com/*",
            },]
        },
    }));

    // snapshot because values depend on CDK implementation
    const tpl = SynthUtils.toCloudFormation(stack);
    const ress = tpl.Resources;
    const res = Object.keys(ress).find(k => k.startsWith("CDN"));
    expect(res).toMatchSnapshot();

});
