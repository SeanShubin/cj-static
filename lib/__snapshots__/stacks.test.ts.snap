// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Certificate Stack Has Certificate Arn Output 1`] = `
Object {
  "Value": Object {
    "Ref": "Certificate4E7ABB08",
  },
}
`;

exports[`Data Stack Has Content Write Access Policy Arn Output 1`] = `
Object {
  "Export": Object {
    "Name": "WriteContentPolicyArn",
  },
  "Value": Object {
    "Ref": "OriginContentWriteAccess40742427",
  },
}
`;

exports[`Distribution stack has distribution 1`] = `"CDNCFDistribution5C555532"`;
